### ISSUE TITLE
AttributeError: 'SparseDtype' object has no attribute 'itemsize' (Support for Pandas SparseArray columns)

### ISSUE THREAD

**Issue by danielchalef** [dask/fastparquet#464]
Fastparquet does not appear to support writing Dask dataframes with SparseArray columns. Doing so fails with:
```
AttributeError: 'SparseDtype' object has no attribute 'itemsize'
```
Pandas: 0.25.1
Dask: 2.4.0
Fastparquet: 0.3.2

**Comment by martindurant** [comment]
@TomAugspurger , can you think of a general way to deal with the various new types, or do we need a special case fir each? Sparse should en/decode very well to parquet, which also stores an array of nulls and the non-null values.

**Comment by TomAugspurger** [comment]
It's probably special cases for now. We have a general issue at https://github.com/pandas-dev/pandas/issues/20612.

Not relevant for fastparquet, but pyarrow has a `__arrow_array__` protocol that arrays can implement to convert themselves to the correct format for serialization.

**Comment by martindurant** [comment]
Using fastparquet to convert from arrow data structures doesn't seem too useful :|

**Comment by TomAugspurger** [comment]
Yeah, I wasn't suggesting that. That's the approach pyarrow is taking, but
I don't think fast parquet defining a protocol is reasonable.

**Comment by martindurant** [comment]
@danielchalef , would you like to have a stab at encoding the sparse time? I don't expect to have time to do this myself in the near future.

**Comment by danielchalef** [comment]
@martindurant I'd love to say "yes", but time and skill may limit how useful I'd be. Do you have a design approach in mind?

**Comment by martindurant** [comment]
It's been a little while since I've delved in the code, but in the column chunk writer, it basically already separates out the set of nulls and the (non-null) values, around [here](https://github.com/dask/fastparquet/blob/master/fastparquet/writer.py#L401). In the sparse case, this is already done. The only tricky thing would be ensuring that the schema gets the right dtype before writing begins.

When reading these, there would be no fundamental difference between columns created from sparse versus any column that happens to have any nulls. The information could be stored in the metadata, though; @TomAugspurger , do you happen to know if there is a pandas metadata update to specify the "logical" type for an extension column like this?

**Comment by TomAugspurger** [comment]
I don't really follow, but in the case of a SparseArray, the metadata would be on SparseDtype, which has

* subtype: the NumPy dtype:
* fill_value: a scalar

And the two arrays to write are

* sp_values : the non-fill_value values
* sp_indicies : the non-fill_value locations

I don't know if there's any need for null handling. Can you write `NaN` in float arrays?

**Comment by martindurant** [comment]
By metadata, I mean the JSON representation that goes in the footer, because the column itself will be indistinguishable from the same dense array
